### PR TITLE
Flow graphs serialization

### PR DESCRIPTION
**Pull request by carolhmj**
# Changes

- This PR adds "serialize" and "parse" functions to the flow graph, which work similarly to the other serialization functions we have in the framework
- It also standardizes the construction of the blocks, with a generic construction parameters interface (which has a name for all blocks), and a generic "configure" function that configures blocks according to its construction parameters. This is used, for example, in the "multiGate" block to configure how many outputs the block has.
- It also adds a getClassName function for all blocks (and I'm very sorry for the size the PR ended up with 😢 )
- It also adds all the blocks to the class register
- Removed the CustomFunctionBlock since it's not on the spec and I was just using it for tests. Instead now I'm using the Log block and mocking the console.log function

# Examples
- Serialize: #KSTQJU#3
- Parse: #CVDC94#2

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14389/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Review by RaananW — COMMENTED** [review]
Sorry, a few other questions that came out over the weekend

Parse should be, in general, a static function. I noticed that you have a class member parse functions. Is that deliberate? does that mean that you need to first create an instance of this class and only then pass the serialized object to parse it?

Also - the idea with making the name a little smaller was not only for the variable name. Changing FlowGraph to FG (for example, or maybe even FlowGraphBlock = FGB) with many nodes will save a lot of space on the serialized JSON.

File: packages/dev/core/src/FlowGraph/flowGraphDataConnection.ts [hunk 1]
```diff
@@ -76,11 +69,21 @@ export class FlowGraphDataConnection<T> extends FlowGraphConnection<FlowGraphBlo
...
+    }
+
+    public serialize(serializationObject: any = {}) {
+        super.serialize(serializationObject);
+        serializationObject.richType = {};
+        this.richType.serialize(serializationObject.richType);
+    }
+
+    public static Parse(serializationObject: any, ownerBlock: FlowGraphBlock): FlowGraphDataConnection<any> {
+        const obj = super.Parse(serializationObject, ownerBlock);
```
Comment (RaananW) [inline packages/dev/core/src/FlowGraph/flowGraphDataConnection.ts:83]: super in static feels wrong. TBH i didn't even know it was working.. Please use the right object reference

File: packages/dev/core/src/FlowGraph/flowGraphVariableDefinitions.ts [hunk 2]
```diff
@@ -13,11 +13,24 @@ export class FlowGraphVariableDefinitions {
...
         return context;
     }
+
+    public serialize(serializationObject: any) {
+        this._definitions.forEach((value, key) => {
+            serializationObject[key] = value;
+        });
+    }
+
+    public parse(serializationObject: any) {
```
Comment (RaananW) [inline packages/dev/core/src/FlowGraph/flowGraphVariableDefinitions.ts:30]: was this missed, or kept like this deliberatly?

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts [hunk 3]
```diff
@@ -1,14 +1,39 @@
...
+        };
+    } else {
+        serializationObject[key] = value;
+    }
+}
+
+function defaultValueParseFunction(key: string, serializationObject: any, scene: Scene) {
+    const value = serializationObject[key];
+    let finalValue;
+    if (value?.className === "Mesh") {
```
Comment (RaananW) [inline packages/dev/core/src/FlowGraph/flowGraphContext.ts:24]: is it only for mesh? or also for InstanceMesh, GroundMesh, AbstractMesh?

**Review by RaananW — COMMENTED** [review]
Just a few questions and small nits, otherwise LGTM

File: packages/dev/core/src/FlowGraph/flowGraphVariableDefinitions.ts (hunk 2 above)
Comment (carolhmj) [reply packages/dev/core/src/FlowGraph/flowGraphVariableDefinitions.ts:30]: It was a miss, sorry. I'll change it.

File: packages/dev/core/src/FlowGraph/flowGraphDataConnection.ts (hunk 1 above)
Comment (carolhmj) [reply packages/dev/core/src/FlowGraph/flowGraphDataConnection.ts:83]: Will do, thanks

File: packages/dev/core/src/FlowGraph/flowGraphContext.ts (hunk 3 above)
Comment (RaananW) [reply packages/dev/core/src/FlowGraph/flowGraphContext.ts:24]: Makes me wonder if we should have helper functions for things like that. (Not for this PR, just a thought).
isMesh, isLight, isTransfornNode ? maybe a part of tools. I'll see what's possible

File: packages/dev/core/test/unit/FlowGraph/flowGraphSerialization.test.ts [hunk 4]
```diff
@@ -18,6 +18,7 @@ import {
 import { FlowGraphConnectionType } from "core/FlowGraph/flowGraphConnection";
 import { FlowGraphDataConnection } from "core/FlowGraph/flowGraphDataConnection";
 import { Vector3 } from "core/Maths";
+import { Mesh } from "core/Meshes";
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/test/unit/FlowGraph/flowGraphSerialization.test.ts:21]: oh, this is a test. that's good.

**Pull request merged**